### ISSUE TITLE
CA1508 (Avoid dead conditional code) false positive

### ISSUE THREAD

**Issue by Neme12** [dotnet/roslyn-analyzers#6908]
### Analyzer

**Diagnostic ID**: [CA1508](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `Avoid dead conditional code`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: SDK 8.0.100-preview.7.23376.3

### Describe the bug

In the following code:
```c#
using Xunit;

public sealed class CollectionsMarshalPolyfillTest
{
    [Fact]
    public void TestAsSpan_Null()
    {
        var span = AsSpan<int>(null);
        Assert.True(span == default);
    }

    public static Span<T> AsSpan<T>(List<T>? list)
    {
        throw new NotImplementedException();
    }
}
```
I get a warning saying that `span == default` is always false, which isn't true.
![image](https://github.com/dotnet/roslyn-analyzers/assets/11444821/a6af15be-1fb5-4e80-a24a-85a19b725e90)

**Comment by mavasani** [comment]
Confirmed that repros on the latest bits as well.
